### DISCUSSION TITLE
Does quic always build on top of Tcp?

### DISCUSSION THREAD

**Discussion by harsh4786** [libp2p/rust-libp2p#4351] [Q&A] [answered]
A noob question but I’m trying to setup quic, and pass it inside my SwarmBuilder, does quic always have to be built over tcp? I’m asking this after looking at an example in the rust libp2p repo

**Comment by mxinden** [comment]
I am not sure what you are referring to with "over". Can you rephrase?

As a good example on how to use QUIC, see:

https://github.com/libp2p/rust-libp2p/blob/608e2dbb198761f5a7e36d05aa27269cc9b6bcb0/examples/chat-example/src/main.rs#L51-L64

Note that this will become easier with https://github.com/libp2p/rust-libp2p/pull/4120.

**Reply by harsh4786** [reply]
By "over" I mean first initializing a TCP transport object and then initializing quic? I mean does it always have to be me initializing TCP before quic or quic can work as a standalone transport?

That's the exact example I was referring to

**Reply by thomaseizinger** [reply]
Nope, you can also just initialize a QUIC transport! Can you post the error you are hitting?

**Reply by harsh4786** [reply]
Getting this error when I try to initialize a QUIC object and pass it inside my SwarmBuilder, I know it's a type issue but how to achieve that type.
<img width="921" alt="Screenshot 2023-08-22 at 12 01 54 AM" src="https://github.com/libp2p/rust-libp2p/assets/50767810/7d27878f-831c-4776-b4dc-436099a8afe5">

**Reply by thomaseizinger** [reply]
You still need to do the `.map` and box up the connection object produced by the QUIC transport. See line 61 in the above example.

**Reply by harsh4786** [reply]
I can't upgrade my quic transport to have `tls`

**Reply by thomaseizinger** [reply, marked as answer]
I am not sure what you mean? QUIC implies TLS, you don't need to explicitly configure it. Can you share the code that is not working?

**Reply by harsh4786** [reply]
Ohh I see, thanks. I thought I had to explicitly add TLS authentication to my quic transport
